src/Native/System.Native/pal_readdir.cpp

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#include "../config.h"
+#include "pal_readdir.h"
+#include <errno.h>
+
+static void ConvertDirent(const dirent& entry, DirectoryEntry* outputEntry)
+{
+    // We use Marshal.PtrToStringAnsi on the managed side, which takes a pointer to
+    // the start of the unmanaged string. Give the caller back a pointer to the 
+    // location of the start of the string that exists in their own byte buffer.
+    outputEntry->nameStartAddress = &entry.d_name[0];
+    outputEntry->inode_type = (NodeType)entry.d_type;
+
+#if HAVE_DIRENT_NAME_LEN
+    outputEntry->nameLength = entry.d_namlen;
+#else
+    outputEntry->nameLength = -1; // sentinel value to mean we have to walk to find the first \0
+#endif
+}
+
+extern "C" 
+int32_t GetDirentSize()
+{
+    // Truncating this should be fine
+    return static_cast<int32_t>(sizeof(dirent));
+}
+
+// To reduce the number of string copies, this function calling pattern works as follows:
+// 1) The managed code calls GetDirentSize() to get the platform-specific 
+//    size of the dirent struct.
+// 2) The managed code creates a byte[] buffer of the size of the native dirent
+//    and passes a pointer to this buffer to this function. 
+// 3) This function passes input byte[] buffer to the OS to fill with dirent data
+//    which makes the 1st strcpy.
+// 4) The ConvertDirent function will set a pointer to the start of the inode name
+//    in the byte[] buffer so the managed code and find it and copy it out of the 
+//    buffer into a managed string that the caller of the framework can use, making
+//    the 2nd and final strcpy. 
+extern "C" 
+int32_t ReadDir(DIR* dir, void* buffer, DirectoryEntry* outputEntry, bool* hasData)
+{
+    int result = 0;
+    errno = 0;
+
+    // On successful cal to readdir_r, temp and &resultEntry should point to the same
+    // data; a NULL temp pointer but return of 0 means that we reached the end of the 
+    // directory stream; finally, a NULL temp pointer with a non-zero return value
+    // means an error occurred.
+    dirent* temp = nullptr;
+    dirent* resultEntry = (dirent*)buffer;
```

Comment:
Not really fundamentially wrong if larger if you take the perspective of documenting the function on its own.  I'd make the call actually fail if it's too small, not just assert. I realize this is academic with only one caller in practice, but I think it's worthwhile to follow the discipline for anything taking in a writable buffer.

I'd doc the bufferSize parameter as "must be at least as large as GetDirentSize() and never needs to be any larger."

Finally, I'd use uint8_t* instead of void* as the type.